index.coffee

```diff
@@ -148,20 +35,171 @@ onProgressHandler = (onProgressPromise, fallbackOnProgress) ->
+						downloadedPercentage = 100
+						extractedPercentage = 100
+					else
+						downloadedPercentage = downloadProgressTracker.getProgress()
+						extractedPercentage = extractionProgressTracker.getProgress()
+
+					percentage = (downloadedPercentage + extractedPercentage) // 2
+					percentage = lastPercentage = Math.max(percentage, lastPercentage)
+
+					onProgress _.merge evt, {
+						percentage
+						downloadedPercentage
+						extractedPercentage
+						totalProgress: progressRenderer(percentage)
+					}
+				catch err
+					console.warn('Progress error:', err.message ? err)
+
+	# Create a stream that transforms `docker.modem.followProgress` onProgress
+	# events to include total progress metrics.
+	pushProgress: (image, onProgress) ->
+		progressRenderer = @renderProgress
+		progressTracker = new ProgressTracker(100 * 1024) # 100 KB
+		lastPercentage = 0
+		Promise.try ->
+			return (evt) ->
+				try
+					{ id, status } = evt
+
+					pushMatch = /Image (.*) already pushed/.exec(status)
+
+					id ?= pushMatch?[1]
+					status ?= ''
+
+					if status is 'Preparing'
+						progressTracker.addLayer(id)
+					else if status is 'Pushing' and evt.progressDetail.current?
+						progressTracker.updateLayer(id, evt.progressDetail)
+					# registry v2 statuses
+					else if _.includes(['Pushed', 'Layer already exists', 'Image already exists'], status) or /^Mounted from /.test(status)
+						progressTracker.finishLayer(id)
+					# registry v1 statuses
+					else if pushMatch? or _.includes(['Already exists', 'Image successfully pushed'], status)
+						progressTracker.finishLayer(id)
+
+					percentage = if status.match(/^latest: digest: /) or status.match(/^Pushing tag for rev /)
+						100
+					else
+						progressTracker.getProgress()
+
+					percentage = lastPercentage = Math.max(percentage, lastPercentage)
+
+					onProgress _.merge evt, {
+						id
+						percentage
+						totalProgress: progressRenderer(percentage)
+					}
+				catch err
+					console.warn('Progress error:', err.message ? err)
 
 exports.DockerProgress = class DockerProgress
 	constructor: (dockerOpts) ->
 		if !(this instanceof DockerProgress)
 			return new DockerProgress(dockerOpts)
 
 		@docker = new Docker(dockerOpts)
+		@reporter = null
+
+	getProgressRenderer: (stepCount = DEFAULT_PROGRESS_BAR_STEP_COUNT) ->
+		return (percentage) ->
+			renderProgress(percentage, stepCount)
+
+	getProgressReporter: ->
+		return @reporter if @reporter?
+		docker = @docker
+		renderer = @getProgressRenderer()
+		@reporter = docker.versionAsync().then (res) ->
+			version = res['Version']
+			if version? and semver.lt(version, LEGACY_DOCKER_VERSION)
+				LegacyProgressReporter = require('./legacy').ProgressReporter
+				return new LegacyProgressReporter(renderer, docker)
+			else
+				return new ProgressReporter(renderer)
+
+	aggregateProgress: (count, onProgress) ->
+		renderer = @getProgressRenderer()
+		states = []
+		reporters = []
+		for i in [0...count]
```

Comment:
Yep, makes sense and that is also more readable.

`states` not having been populated isn't really an issue though, because the returned progress handler isn't executed until much later (it's simply bound at the time but not used) so there's no risk of invoking the callback with partially calculated results.

Also `combinedProgressHandler` could actually be defined inside `aggregateProgress` guaranteeing that it won't be re-used in any other context, it just feels better to be beside the normal progress handler (even though with `_.times` it's now a lot more tempting to move it inside).